### ISSUE TITLE
Why is non-const& parameter allowed in `ranges::view::transform` function?

### ISSUE THREAD

**Issue by denchat** [ericniebler/range-v3#798]
Shouldn't `ranges::view::transform` prohibit users from modify underlining container?
```
    auto vec = std::vector{ 200, 300, 400, 500 };

    auto plus_one_by_ref = [](auto& n) { return n++ + 10; }; // modify element by ref
    auto rng_plus_ten_but_secretly_plus_one = vec | ranges::view::transform(plus_one_by_ref);
    
    // activate the laziness twice
    ranges::for_each(rng_plus_ten_but_secretly_plus_one, [](const auto& in){});
    ranges::for_each(rng_plus_ten_but_secretly_plus_one, [](const auto& in){});

    // print [202,302,402,502]
    std::cout << (vec | ranges::view::all) << "\n";
```
[godbolt.org/g/gLy2Ap](https://godbolt.org/g/gLy2Ap)

**Comment by ericniebler** [comment]
> Shouldn't `ranges::view::transform` prohibit users from modify underlining container?

Prohibiting transformation functions that accept arguments by non-`const` `&` feels like excessive nannying to me. The case can be made that the transformation function should satisfy `RegularInvocable` instead of `Invocable`, but that would prohibit transformation functions that maintain internal state, which is useful. I don't think there is anything to be done here.

@CaseyCarter, care to proffer your opinion here?

**Comment by denchat** [comment]
Perhaps we can try keeping the premise just by naming like:

`range::view::transform(RegularInvocable f)`
`range::mutable::transform(Invocable f)`

or even just the other way around :

`range::const::transform(RegularInvocable f)`
`range::view::transform(Invocable f)`

Maybe this can improve some aspect of 'harder to wrongly use'.

**Comment by ericniebler** [comment]
I don't feel that this is a problem that needs fixing. Did this lead design to an actual but in your code, are is your concern purely hypothetical? Nobody else has reported problems with this design.

Closing, but feel free to follow up.

**Comment by CaseyCarter** [comment]
> @CaseyCarter, care to proffer your opinion here?

Yes, we should prohibit users from modifying the element passed to the transform function (modifying the other elements is already forbidden by the blanket "value stability" wording). If the transform function modifies the element, then `transform_view` as specified in P0789 isn't a view because its iterators don't satisfy `Readable` (`*i` for a `Readable` `i` is VERY indirectly required to be equality-preserving).

I don't agree that forcing arguments to `const&` is necessary or sufficient to achieve that goal. Especially in the world of proxy iterators, forcing functions to accept the result of iterator dereference by `const&` doesn't do much. Even absent proxy iterators, `mutable` and `const_cast` exist in the IS and are perfectly valid. 

I *do* agree that despite that we cannot make it impossible for users to modify things inappropriately, it behooves us to make it harder to misuse the library. (Murphy vs. Machiavelli, etc.) I have been entertaining a notion of a customization point for iterators that would be used to examine the element the iterator denotes in a manner that is guaranteed (by specification) to be non-modifying, which would best-effort "constify" the returned reference. ([LWG3031](https://cplusplus.github.io/LWG/issue3031) has provided some impetus for this, and discussions with Marshall Clow prompted by my PR for [LWG3054](https://cplusplus.github.io/LWG/issue3054) provided the rest.)

```c++
// Pretend these two functions are fallback implementations of a proper Customization Point Object definition
template<class R>
requires Readable<remove_reference_t<R>>
decltype(auto) iter_dont_move(R&& r) // intentionally horrible strawman name
{
  return *std::forward<R>(r);
}

template<class R>
requires Readable<remove_reference_t<R>> && 
  std::is_lvalue_reference_v<reference_t<R>> &&
  Same<remove_cvref_t<reference_t<R>>, value_type_t<R>>
add_const_t<reference_t<R>> iter_dont_move(R&& r)
{
  return *std::forward<R>(r);
}
```
Proxy iterators would hook the customization so that `iter_dont_move(i)` returns something "proxy reference to `const`"-ish when simply `*i` would return something "proxy reference to mutable-ish". Algorithms would use the CP when performing examination:
```c++
template<InputIterator I, Sentinel<I> S, class F>
requires /* .... constrain F ... */
void copy_if(I first, S const last, F f) {
  for (l first != last; ++first) {
    if (f(iter_dont_move(first))) { // Forbidden to modify *first, e.g. by moving from
      *last++ = *first; // Allowed to modify *first, e.g. by moving from
    }
  }
}
```
To avoid inadvertent modification, such is in the presence of `move_iterator`s. This would notably make it possible to implement multi-pass algorithms that work with xvalue (or xvalue-like proxy) iterators even when those algorithms create local variables.

Blanket wording would specify that "required expressions that include a subexpression `iter_dont_move(E)` shall not modify the value denoted by `E`" so we need not repeat it every for every use.

`iter_dont_move` would also be used to implement `view::const`, which will otherwise be quite challenging in proxy iterator world.

TLDR: I've been thinking about this, but I need to find a few free days to experiment with implementation  and write up a proposal.

**Comment by CaseyCarter** [comment]
I don't disagree. I do think the investigation itself would be valuable, however. Much like particle accelerator research, every idea we fire at the proxy reference problem teaches us a little more about it.
